AC-223 updating transcript skip links

clrux opened the pull request:
This work relates to [AC-223](https://openedx.atlassian.net/browse/AC-223) and addresses broken page anchor links on pages with video player modules. The links are meant to quickly navigate to the start and end of the transcript (if present), but weren't skipping anywhere.

## Note on accessibility errors

One of the accessibility failures references an `<a href="#">` that's used as the drag handle on the video slider. This is created by the version of jQuery UI we're using in the platform. I believe more recent versions of jQuery UI use better markup for this.

The other two should not exist as the `href` points to an `id` that's built when the video is inserted. I'm not sure if the test runs before the video has a chance to load, but those page anchors do in fact work.

## Sandbox

https://clrux-ac-223.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/d8a6192ade314473a78242dfeedfbf5b/edx_introduction/
https://studio-clrux-ac-223.sandbox.edx.org

## Reviewers 

- [x] @clytwynec
- [x] T&L 2nd review

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -82,8 +82,15 @@
                     /* jshint maxlen:200 */
                     gettext('Activating an item in this group will spool the video to the corresponding time point. To skip transcript, go to previous item.'),
                     '">',
+                    '<h3 id="transcript-label" class="sr">Video transcript</h3>',
```
You should go ahead and translate text for sr classes (use gettext).

clytwynec commented:
@clrux  Have you tried un-ignoring the tests that were failing from this to see if they pass now? ([test_studio_video_module.py](https://github.com/edx/edx-platform/blob/a6e838348942856ebd5044f13e41ef679cd0056e/common/test/acceptance/tests/video/test_studio_video_module.py#L348) and [test_video_module.py](https://github.com/edx/edx-platform/blob/d6673ca3fd5104518f07c0006758b0e14db6e500/common/test/acceptance/tests/video/test_video_module.py#L1211))

cahrens commented:
I think that duplicate IDs can't be introduced. You can generate a unique in JS (see usages of _.uniqueId()) and pass it to the template, or find a way to pass through the xblock usage ID.

clrux commented:
jenkins run bokchoy

clrux commented:
@clytwynec @cahrens please see my note in the description as to why I can't get `a11y` tests to pass.

clrux commented:
jenkins run bokchoy
jenkins run js

clytwynec commented:
@clrux  You should be able to confirm whether or not the video is loaded at the time the test runs.  It may also be that the the targets are hidden at the time the tests run.  

For the drag handles (and possible the other failing links), you can specifically ignore those elements in the test.  That way the rule will still run agains the rest of the page.  See the [Bok Choy documentation](http://bok-choy.readthedocs.org/en/latest/api_reference.html#bok_choy.a11y.axe_core_ruleset.AxeCoreAuditConfig.set_scope).

clrux commented:
jenkins run bokchoy
jenkins run js

clrux commented:
jenkins run bokchoy

clrux commented:
jenkins run bokchoy

clrux commented:
jenkins run bokchoy

clrux commented:
jenkins run lettuce

clrux commented:
@clytwynec this is ready for your review. I'm hoping you can take a look at the tests?
@cahrens this is ready for your review. I'm hoping you can take a look at the overall JS?

cahrens commented:
@clrux and @clytwynec Please have Christine do a complete review of the PR first, not just the tests. We are still moving towards the goal of having TNL just do a quick 2nd review after a complete first review is done.

cahrens commented:
TNL review ticket-- https://openedx.atlassian.net/browse/TNL-4162

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -117,10 +74,62 @@
             renderElements: function () {
                 var languages = this.state.config.transcriptLanguages;
 
+                var langTemplate = [
```
@clrux  What is the reason to move `langTemplate` and `template` into `renderElements`?   It seems like an unnecessary change to me (but there may be some convention I'm not aware of), and makes it difficult to figure out what actually changed in the templates.

clrux replied:
I needed to be able to access variables, which wasn't possible the other way. I spoke with Andy A about this change and he was OK with it.

clytwynec replied:
@clrux  Could you make it a real template and then render it in this context instead?  That would allow you to use the variables in the template without having it as part of this function's logic.

clrux replied:
This method follows that of a similar method above. If I changed the way this one template works I'd want to change the other, but I'm thinking that's for another PR. Thoughts?

clytwynec replied:
Where is the other method you're referring to?

clrux replied:
`langTemplate` was already in this file. I added `template`.

clytwynec replied:
I'm of the opinion that these should be templates with proper escaping.

clrux replied:
Is there an example of this that I can follow?

clytwynec replied:
I'd take a look at the [underscore docs](http://underscorejs.org/#template).

clrux replied:
I've moved this work to another PR for future work. It's a bit out of scope for this. This work only includes fixing the skip links to/within the transcript.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -553,11 +562,6 @@
                         self.loaded = true;
                     },
                     error: function (jqXHR, textStatus, errorThrown) {
-                        console.log('[Video info]: ERROR while fetching captions.');
```
These logs seem like they might be useful.  Are you sure we should be removing them?

clytwynec commented on common/test/acceptance/pages/lms/video/video.py:
```diff
@@ -57,7 +57,8 @@
     'download_transcript': '.video-tracks .a11y-menu-list',
-    'transcript-format': '.video-tracks .a11y-menu-button'
+    'transcript-format': '.video-tracks .a11y-menu-button',
+    'transcript-skip': '.sr-is-focusable.transcript-start'
```
I'd add a comma at the end of the last line in this dict. (See [why-does-python-allow-commas-at-the-end-of-lists-and-tuples](https://docs.python.org/2/faq/design.html#why-does-python-allow-commas-at-the-end-of-lists-and-tuples).)

clytwynec commented on common/test/acceptance/pages/studio/video/video.py:
```diff
@@ -48,6 +48,10 @@
 }
 
+VIDEO_MENUS = {
+    'transcript-skip': '.sr-is-focusable.transcript-start'
```
Same here. I'd add a comma at the end of the last (only) line in this dict. (See [why-does-python-allow-commas-at-the-end-of-lists-and-tuples](https://docs.python.org/2/faq/design.html#why-does-python-allow-commas-at-the-end-of-lists-and-tuples).)

clytwynec commented on common/test/acceptance/pages/lms/video/video.py:
```diff
@@ -390,6 +391,22 @@ def captions_text(self):
         return ' '.join(subs)
 
     @property
+    def captions_container(self):
```
Where is this new property used?

clrux replied:
Ah, looks like the remnants of a previously tested method. Removing.

clytwynec commented on common/test/acceptance/tests/video/test_studio_video_module.py:
```diff
@@ -341,13 +341,18 @@ def setUp(self):
         self.outline.a11y_audit.config.set_rules({
             "ignore": [
-                'link-href',  # TODO: AC-223
+                # 'link-href',  # TODO: AC-223
```
Since this isn't ignoring anything anymore, the entire `self.outline.a11y_audit.config.set_rules` call should be removed here.

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -1205,13 +1205,21 @@ def setUp(self):
+            include=["div.video"],
+            exclude=["a.ui-slider-handle"]
+        )
         self.video.a11y_audit.config.set_rules({
```
Same thing here.  Since it's no longer really doing anything, just remove the whole `self.video.a11y_audit.config.set_rules` call.

clytwynec commented on lms/templates/video.html:
```diff
@@ -33,12 +32,10 @@ <h4 class="video-error hidden">${_('No playable video sources found.')}</h4>
                   <div class="secondary-controls"></div>
               </div>
           </section>
-          <a class="nav-skip sr" id="before-transcript_${id}" href="#after-transcript_${id}">${_('Skip to end of transcript.')}</a>
```
@clrux  Does this mean that there is no longer the capability to skip to the end of a transcript?   What will the impact of that be on users?

clrux replied:
I've added the links to skip back and forth between the start and end of the transcript.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/01_initialize.js:
```diff
@@ -131,16 +131,12 @@ function (VideoPlayer, i18n, moment) {
             state.youtubeApiAvailable = false;
 
             onYTApiReady = function () {
-                console.log('[Video info]: YouTube API is available and is loaded.');
```
I have the same concern here regarding the removal of log info.  These seem like they were intentionally put here and removing them is unrelated to the other work in this PR.  Is it possible we are somehow tracking this logged information for certain use cases?

clytwynec commented:
@clrux  I'm done with my first pass at this.  I had some questions. In particular, one about how removing the option to skip to the end of the transcript will impact users.

clrux commented:
I've removed the two skip links that appear before and after the video, but added one at the top of the transcript that skips to the end, and one at the end that skips to the start of the transcript. This should provide solid functionality for both keyboard and screenreader users.

@clytwynec I've also replaced the previously removed log outputs. I removed them because they were making the console very noisy.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -117,10 +74,63 @@
+                        'corresponding time point.'),
+                    '" id="transcript-' + this.state.id + '">',
+                        '<a href="#transcript-end-' + this.state.id + '" id="transcript-start-'
+                        + this.state.id + '">Start of transcript. Skip to the end.</a>',
```
The text in this link should be internationalized.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -117,10 +74,63 @@
+                            gettext('Video transcript'),
+                        '</h3>',
+                        '<ol id="transcript-captions" class="subtitles-menu"></ol>',
+                        '<a href="#transcript-start-' + this.state.id + '" id="transcript-end-'
```
Same here. The text in this link should be internationalized.

clrux commented:
jenkins run bokchoy

clrux commented:
@clytwynec I've moved the inline templates to Underscore. Please take a second pass when you get a chance.

@andy-armstrong I've tagged you to review my newbie Underscore usage. We wanted to clean up the code a bit so I moved the concatenated templates to Underscore. Please review this PR at your earliest convenience.

clrux commented:
@clytwynec and @andy-armstrong hold your reviews, I found an issue. I'll ping you when it's ready.

clrux commented:
@clytwynec please continue your review. @andy-armstrong when Christine finishes, would you mind taking the T&L second pass?

cahrens commented:
@clrux We have a TNL review ticket for this already. It will be done by me, Diana, or Eric (@andy-armstrong is working on FEDX right now, not TNL). If you have specific questions for Andy, tag him just on those portions of the PR.

clrux commented:
@cahrens Cool, thanks. I'll update the description.

clytwynec commented:
@clrux just a couple of questions. But :+1: once they are addressed.

clrux commented:
@clytwynec For some reason the jshint maxlen wasn't working so I removed it and kept the concatenated string. Is this okay?

clytwynec commented:
@clrux  That should be fine as long as you've manually confirmed that the translations work properly for that string.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -117,10 +74,62 @@
+
+                var template = [
+                    '<div class="subtitles" role="region" aria-label="',
+                        gettext('Video transcript: activating an item in this group will spool the video ',
```
You can't break up JS strings into multiple lines and have it work with gettext. Please use single-line strings and also test that translations are working properly.

See http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#don-t-append-strings-interpolate-values

And
http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#coverage-testing

clrux commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -546,6 +539,28 @@
+                                    .attr('aria-label', gettext('Language: Press the UP arrow key to enter the language menu then use UP and DOWN arrow keys to navigate language options. Press ENTER to change to the selected language.'))
+                                    .find('.control-text')
+                                        .text(gettext('Open langage menu.'));
                         }
```
The i18n didn't want to work in the earlier object that was $.join'ed, but after looking at how other files handle this, I've moved these here to be updated later.

cahrens commented:
@clrux Please let me know when this PR is ready for TNL review. I'm not clear where you are with the i18n fixes and all.

clrux commented:
@cahrens Any ideas why `jshint maxlen` wouldn't be working here?

cahrens commented:
@clrux Hmmm, not sure. Perhaps because of the placement of the periods (next line instead of same line)? I personally don't like that style, so I'd like to think that it is wrong. :)

clrux commented:
@andy-armstrong I've got two troublesome lines that are too long. Unfortunately they're `gettext` strings that can't be broken into multiple lines, but they both trigger line length failures in JSHint. I was looking at `maxlen` and it looks like it's going to be deprecated in the future, so how do you advise I handle these two lines?

andy-armstrong commented:
@clrux I've been adding the following to the end of long lines:

    // jshint ignore:line

See here, for example:

https://github.com/edx/edx-platform/blob/andya/add-ui-toolkit/lms/djangoapps/teams/static/teams/js/spec/views/team_discussion_spec.js#L57

clrux commented:
Thanks @andy-armstrong

clrux commented:
jenkins run js

clytwynec commented on common/test/acceptance/pages/studio/video/video.py:
```diff
@@ -48,6 +48,10 @@
     'collapse_link': '.collapse-action.collapse-setting',
 }
 
+VIDEO_MENUS = {
```
Why is this needed here?

clrux replied:
Remnants from failed experiments. Removing.

clytwynec commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -37,8 +37,8 @@ msgid ""
 msgstr ""
```
These files (the `.po` files modified) shouldn't be checked in. From the [open edx internationalization docs](http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html#coverage-testing):  `This step creates new dummy translations in the Esperanto directory (edx-platform/conf/local/eo/LC_MESSAGES) and the RTL directory (edx-platform/conf/local/rtl/LC_MESSAGES). DO NOT CHECK THESE FILES IN. You should discard these files once you have finished testing.`

clytwynec commented on lms/static/js/i18n/eo/djangojs.js:
```diff
@@ -137,7 +137,7 @@
     "<%= user %> has been successfully added to the exception list. Click Generate Exception Certificate below to send the certificate.": "<%= user %> h\u00e4s \u00df\u00e9\u00e9n s\u00fc\u00e7\u00e7\u00e9ssf\u00fcll\u00fd \u00e4dd\u00e9d t\u00f6 th\u00e9 \u00e9x\u00e7\u00e9pt\u00ef\u00f6n l\u00efst. \u00c7l\u00ef\u00e7k G\u00e9n\u00e9r\u00e4t\u00e9 \u00c9x\u00e7\u00e9pt\u00ef\u00f6n \u00c7\u00e9 ...
```
I'm doubting that these files are meant to be checked in too but I'm not sure....  Maybe @dianakhuang or @nedbat  knows?

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -536,10 +529,7 @@
-                                        '<li>Transcript will be displayed when ' +
-                                        'you start playing the video.</li>'
-                                    )
+                                    gettext('<li>Transcript will be displayed when you start playing the video.</li>')
```
The html shouldn't be included as part of the translation.   A translator may not know what it is and it may not be kept, or could be mistranslated.

andy-armstrong replied:
Good catch @clytwynec. We should update this to follow the safe templates best practice: https://openedx.atlassian.net/wiki/display/TNL/Safe+Template+Marathon

clytwynec commented:
I had a couple of questions about the translations, but once those are resolved 👍

clrux commented:
@andy-armstrong Mind reviewing this one from T&L?

clrux commented:
@cahrens This is ready for you review. Can I get an estimate?

cahrens commented:
@clrux Sure, we can get to this early next week.

clrux commented:
@cahrens Thank you ma'am. I'll follow up next week!

dianakhuang commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -747,6 +737,21 @@
                     self.rendered = true;
                     self.state.el.addClass('is-captions-rendered');
+
+                    self.subtitlesEl.attr('aria-label', gettext('Video transcript: activating an item in this group will spool the video to the corresponding time point.')) // jshint ignore:line
```
Has someone from docs reviewed the text? I understand what 'spool' means in this context, but it's odd phrasing to me.

clrux replied:
@lamagnifica would you mind offering a suggestion on better phrasing here?

lamagnifica replied:
@clrux happy to contribute. Could you give me a little more context? I read the PR and JIRA story, and tried to look at both sandboxes (not found), but I'm not clear on what exactly the learner is doing when this message appears.

clrux replied:
This text is for screenreader only users and it describes how the transcript links work. Activating a link in the transcript will skip to that point in time in the video. I believe "spool" was here originally, but I did some shuffling with the code so it appears new.

dianakhuang commented:
Other than the questions about user-visible strings that I'll leave up to @lamagnifica , :+1: from me.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -536,10 +529,7 @@
-                                        '<li>Transcript will be displayed when ' +
-                                        'you start playing the video.</li>'
-                                    )
+                                    '<li>' + gettext('Transcript will be displayed when you start playing the video.') + '</li>' // jshint ignore:line
```
Yeah, the concatenation of HTML and translated text needs to be done with HTMLUtils per the safe template best practices. @clrux this is being done in a different PR?

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -747,6 +737,21 @@
+                        .find('.transcript-title')
+                            .html(gettext('Video transcript'));
+
+                    self.subtitlesEl.find('.transcript-start')
```
@clrux Since this portion of the code is new, I do think you need to be escaping the text being set here (in case someone translates these strings to have characters that need to be escaped). I believe you want to call the jquery "text" method instead of "html" method. 

http://api.jquery.com/text/

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -747,6 +737,21 @@
                     self.isMouseFocus = false;
                     self.rendered = true;
                     self.state.el.addClass('is-captions-rendered');
+
```
This "attr" value also needs to be escaped (because you don't know what may be in the translation). @robrap is the best practice here just to use _.escape?

clrux replied:
It looks like `attr()` is automatically made safe by jQuery.

clrux replied:
It looks like `attr()` from jQuery is escape automatically?

"The final step of DOM manipulation in JavaScript often happens using JQuery. There are some JQuery functions such as $.text(), $.attr() and $.val() that expect plain text strings and take care of HTML-escaping its input for you."

cahrens replied:
OK, I didn't see any mention of it here: http://api.jquery.com/attr/

cahrens commented:
@clrux I do think that proper escaping should be done on the new code-- I added some notes.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -747,6 +749,43 @@
+                    self.subtitlesEl
+                        .attr('aria-label', gettext('Video transcript: activating an item in this group will spool the video to the corresponding time point.')); // jshint ignore:line
+
+                    HtmlUtils.setHtml(
```
@clrux when you are setting text that is not HTML, I confirmed with @andy-armstrong that the best practice is to use JQuery's .text method (instead of HtmlUtils.setHtml).

clrux commented:
@cahrens I'm having RequireJS issues when loading `backbone`, `underscore`, `gettext`, and `HtmlUtils` as they aren't found. I checked other files that use the first three of these and they are loaded the same. I switched to master to see if this was specific to my branch, but I get the same error on master. Either they're being loaded improperly, or there's something wrong with the dependencies.

Note that during the testing, console logs report they can't find the files, as such:
`/suite/backbone.js`, `/suite/underscore.js`, and so on. Although these are defined with different paths in `common.py`, etc., the paths look incorrect, which may be the issue.

Do you or @andy-armstrong have thoughts on this?

clrux commented:
jenkins run js

clrux commented:
jenkins run js

clrux commented:
paver run python

clrux commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -551,7 +529,9 @@
                         } else {
                             if (state.isTouch) {
-                                this.showTemporaryCaptionMessage(self, "<li>Transcript will be displayed when you start playing the video.</li>"); // jshint ignore: line
+                                self.subtitlesEl.find('.subtitles-menu')
```
@cahrens or @andy-armstrong I was unable to get HtmlUtils working in the tests (it's available when previewing locally) so I did some research an opted for this method. It uses `$.text()` then `$.wrap()` so I don't see any problems with escaping. Is this OK?

cahrens replied:
If the functionality is working properly, this is safe from an XSS perspective.

clrux commented:
jenkins run python

clrux commented:
jenkins run python

clrux commented:
@cahrens This is once again ready for you to continue your review.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -4,7 +4,7 @@
     define('video/09_video_caption.js', [
         'video/00_sjson.js',
-        'video/00_async_process.js'
+        'video/00_async_process.js',
```
No trailing commas in JavaScript files, please.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -759,8 +739,6 @@
                     self.rendered = true;
                     self.state.el.addClass('is-captions-rendered');
 
-                    self.showTemporaryCaptionMessage(self, "<li>Transcript will be displayed when you start playing the video.</li>"); // jshint ignore: line
```
Why is this line not necessary anymore?

clrux replied:
I created that function originally to run through the HtmlUtil stuff in case these messages were used more than once. Turns out this is the only place so I simplified it and did it inline, removing the function since it wasn't being used.

cahrens commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -1101,14 +1110,14 @@
                 } else {
                     this.captionDisplayEl
-                        .text(gettext('(Caption will be displayed when you start playing the video.)'));
+                        .text(window.gettext('(Caption will be displayed when you start playing the video.)'));
```
Nit: you could have left all of these as just gettext. I agree that the file should be consistent, so if you back out these changes, also change your new usages to just be gettext.

cahrens commented:
👍 on XSS concerns. Note that I have NOT tested the actual functionality.

@clrux I am closing out the TNL review ticket for this PR.
